zebra-state/src/service.rs

```diff
@@ -73,14 +72,33 @@ impl StateService {
     /// in RFC0005.
     ///
     /// [1]: https://zebra.zfnd.org/dev/rfcs/0005-state-updates.html#committing-non-finalized-blocks
-    #[instrument(skip(self, new))]
-    fn queue_and_commit_non_finalized_blocks(&mut self, new: QueuedBlock) {
-        let parent_hash = new.block.header.previous_block_hash;
+    #[instrument(skip(self, block))]
+    fn queue_and_commit_non_finalized_blocks(
+        &mut self,
+        block: Arc<Block>,
+    ) -> oneshot::Receiver<Result<block::Hash, BoxError>> {
+        let hash = block.hash();
+        let parent_hash = block.header.previous_block_hash;
 
-        self.queued_blocks.queue(new);
+        if self.contains_committed_block(&block) {
+            let (rsp_tx, rsp_rx) = oneshot::channel();
+            let _ = rsp_tx.send(Err("duplicate block".into()));
+            return rsp_rx;
+        }
+
```

Comment:
It would be nice to have a comment here explaining the queue of blocks to be verified is a pipeline and that replacing older queued requests by newer ones is a way to lazily flush the pipeline.